reference/repository.go

```diff
@@ -0,0 +1,136 @@
+package reference
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+const (
+	// RepositoryNameTotalLengthMax is the maximum total number of characters in a repository name.
+	RepositoryNameTotalLengthMax = 255
+)
+
+// RepositoryNameComponentRegexp restricts registry path component names to
+// start with at least one letter or number, with following parts able to
+// be separated by one period, dash or underscore.
+var RepositoryNameComponentRegexp = regexp.MustCompile(`[a-zA-Z0-9]+(?:[._-][a-z0-9]+)*`)
+
+// RepositoryNameComponentAnchoredRegexp is the version of
+// RepositoryNameComponentRegexp which must completely match the content
+var RepositoryNameComponentAnchoredRegexp = regexp.MustCompile(`^` + RepositoryNameComponentRegexp.String() + `$`)
+
+// RepositoryNameHostnameRegexp restricts the registry hostname component of a repository name to
+// start with a component as defined by RepositoryNameComponentRegexp and followed by an optional port.
+var RepositoryNameHostnameRegexp = regexp.MustCompile(RepositoryNameComponentRegexp.String() + `(?::[0-9]+)?`)
+
+// RepositoryNameHostnameAnchoredRegexp is the version of
+// RepositoryNameHostnameRegexp which must completely match the content.
+var RepositoryNameHostnameAnchoredRegexp = regexp.MustCompile(`^` + RepositoryNameHostnameRegexp.String() + `$`)
+
+// RepositoryNameRegexp builds on RepositoryNameComponentRegexp to allow
+// multiple path components, separated by a forward slash.
+var RepositoryNameRegexp = regexp.MustCompile(`(?:` + RepositoryNameHostnameRegexp.String() + `/` + RepositoryNameComponentRegexp.String() + `/)*` + RepositoryNameComponentRegexp.String())
```

Comment:
These regexps seem to behave very differently from the existing regexps in api/v2. I tried to use this RepositoryNameRegexp in notary and ran into some problems.

It looks like RepositoryNameRegexp is set up to allow multiple "hostname/component" matches instead of multiple components. When I tested the regexp with regexr.com, I found that:

* docker.com/notary does not match
* docker.com/notary/notary does match
* docker.com/notary/notary/notary does not match

The exact regexp I tested was `(?:[a-zA-Z0-9]+(?:[._-][a-z0-9]+)*(?::[0-9]+)?/[a-zA-Z0-9]+(?:[._-][a-z0-9]+)*/)*[a-zA-Z0-9]+(?:[._-][a-z0-9]+)*`